Add ApptainerWorkspace implementation for rootless container support

neubig opened the pull request:
HUMAN: this has been tested

## Description

This PR implements `ApptainerWorkspace`, a container-based workspace that uses Apptainer (formerly Singularity) instead of Docker. This addresses the need for rootless container execution in HPC and shared computing environments where Docker may not be available or permitted.

**✨ Critical Bug Fix (2025-10-24):** Discovered and fixed a bug where the initial implementation incorrectly used `apptainer build ... docker-daemon://image`, which required Docker to be running. This defeated the entire purpose of Apptainer! The fix changes to `apptainer pull docker://image` which pulls directly from Docker registries without needing Docker daemon. This is the key feature that makes Apptainer valuable.

**✨ Additional Fixes (2025-10-24):**
- **Switched to exec mode**: Changed from Apptainer instance mode to direct `apptainer exec` for better compatibility in environments without systemd/FUSE
- **Fixed authentication**: RemoteWorkspace now properly includes API key in default HTTP client headers, fixing 401 errors when creating conversations
- **Environment variable support**: ApptainerWorkspace now reads `SESSION_API_KEY` from environment and passes it to RemoteWorkspace

Fixes #891

## Key Features

- ✅ **No root privileges required** for container execution
- ✅ **No Docker daemon required** - pulls directly from registries using `apptainer pull`
- ✅ **Works without systemd/FUSE** - uses direct exec mode instead of instances
- ✅ **Converts Docker images to Apptainer SIF format** with intelligent caching
- ✅ **Full RemoteWorkspace API compatibility** - drop-in replacement for DockerWorkspace
- ✅ **Automatic port management** and health checking
- ✅ **Directory mounting** and environment variable forwarding
- ✅ **Proper authentication** via SESSION_API_KEY
- ✅ **Comprehensive documentation** and usage examples
- ✅ **Actually tested end-to-end with Apptainer 1.3.5** - see demo log

## Implementation Details

### Files Added

1. **`openhands-workspace/openhands/workspace/apptainer/workspace.py`** (378 lines)
   - Main `ApptainerWorkspace` class implementation
   - Image preparation using native `apptainer pull` (no Docker required!)
   - Container lifecycle management using `apptainer exec`
   - Health checking and logging
   - Authentication support via SESSION_API_KEY

2. **`openhands-workspace/openhands/workspace/apptainer/__init__.py`**
   - Module initialization and exports

3. **`openhands-workspace/openhands/workspace/apptainer/README.md`**
   - Comprehensive documentation
   - Usage examples for all three image source options
   - Configuration reference
   - Troubleshooting guide

4. **`examples/02_remote_agent_server/05_convo_with_apptainer_sandboxed_server.py`**
   - Complete working example
   - Demonstrates workspace setup, agent conversations, and cleanup

5. **`tests/workspace/test_apptainer_workspace.py`**
   - Import and inheritance tests
   - Field definition validation
   - All tests passing ✅

6. **`apptainer_workspace_demo.log`**
   - Complete log showing successful end-to-end operation
   - Demonstrates container startup, health checks, bash commands, and conversation creation
   - Shows authentication working correctly

### Files Modified

- **`openhands-workspace/openhands/workspace/__init__.py`**
  - Added `ApptainerWorkspace` to exports
- **`openhands-workspace/openhands/workspace/apptainer/workspace.py`**
  - Fixed to use `apptainer pull` instead of Docker daemon
  - Switched to `apptainer exec` instead of instance mode for better compatibility
  - Added SESSION_API_KEY environment variable support
- **`openhands-sdk/openhands/sdk/workspace/remote/base.py`**
  - Fixed to include API key headers in HTTP client initialization

## Usage

### Option 1: Pre-built Server Image (Recommended for HPC)
```python
from openhands.workspace import ApptainerWorkspace

# Pulls from Docker Hub/GHCR without needing Docker!
with ApptainerWorkspace(
    server_image="ghcr.io/openhands/agent-server:main-python",
    host_port=8010,
) as workspace:
    result = workspace.execute_command("echo 'Hello from Apptainer!'")
    print(result.stdout)
```

### Option 2: Build from Base Image (Requires Docker for initial build)
```python
# Note: This still requires Docker to build the agent-server initially,
# but can then be used on systems without Docker
with ApptainerWorkspace(
    base_image="nikolaik/python-nodejs:python3.12-nodejs22",
    host_port=8010,
) as workspace:
    result = workspace.execute_command("python --version")
    print(result.stdout)
```

### Option 3: Use Existing SIF File
```python
with ApptainerWorkspace(
    sif_file="/path/to/agent-server.sif",
    host_port=8010,
) as workspace:
    result = workspace.execute_command("ls -la")
    print(result.stdout)
```

## Testing

All tests pass successfully:

```bash
$ uv run pytest tests/workspace/test_apptainer_workspace.py -v

tests/workspace/test_apptainer_workspace.py::test_apptainer_workspace_import PASSED [ 33%]
tests/workspace/test_apptainer_workspace.py::test_apptainer_workspace_inheritance PASSED [ 66%]
tests/workspace/test_apptainer_workspace.py::test_apptainer_workspace_field_definitions PASSED [100%]

============================== 3 passed in 0.13s ===============================
```

### End-to-End Testing with Actual Apptainer

Successfully tested the complete example with Apptainer 1.3.5. See `apptainer_workspace_demo.log` for full details:

✅ **Image Preparation**
- Successfully pulled Docker image without Docker daemon
- Created SIF file: `/root/.apptainer_cache/ghcr.io_openhands_agent-server_main-python.sif`
- Cached for future use

✅ **Container Execution**
- Container starts successfully using `apptainer exec` mode
- Server starts and listens on port 8010
- Health endpoint responds correctly (HTTP 200)

✅ **Command Execution**
```
Command 'echo 'Hello from Apptainer sandboxed environment!' && pwd' completed with exit code 0
Output: Hello from Apptainer sandboxed environment!
/workspace
```

✅ **Authentication**
- SESSION_API_KEY properly passed to RemoteWorkspace
- API key included in HTTP client headers
- Conversation created successfully (no 401 errors)

✅ **API Endpoints**
- `/health` endpoint: ✅ Working
- `/api/bash/start_bash_command`: ✅ Working
- `/api/conversations`: ✅ Working (with auth)
- `/api/conversations/{id}/run`: ✅ Working (with auth)

All pre-commit hooks pass:
- ✅ Ruff format
- ✅ Ruff lint
- ✅ PEP8 style check (pycodestyle)
- ✅ Type check with basedpyright

## Comparison: ApptainerWorkspace vs DockerWorkspace

| Feature | DockerWorkspace | ApptainerWorkspace |
|---------|----------------|------------------|
| Root privileges | Required (usually) | **Not required** |
| Docker daemon | Required | **Not required** ✨ |
| systemd/FUSE | Not needed | **Not needed** ✨ |
| Container runtime | Docker | Apptainer |
| Image format | Docker images | SIF (from Docker) |
| Use case | General development | **HPC, shared systems** |
| Port mapping | Native | Host networking |
| Exec mode | Process model | **Direct exec** |

## Prerequisites

Users need to install Apptainer: https://apptainer.org/docs/user/main/quick_start.html

On Ubuntu/Debian:
```bash
sudo apt-get install -y apptainer
```

Or build from source:
```bash
wget https://github.com/apptainer/apptainer/releases/download/v1.3.5/apptainer-1.3.5.tar.gz
tar xzf apptainer-1.3.5.tar.gz
cd apptainer-1.3.5
./mconfig && make -C builddir && sudo make -C builddir install
```

## Why Apptainer?

As mentioned in issue #891, Docker requires root access which is often not available or permitted in:
- High-Performance Computing (HPC) environments
- Shared multi-user systems
- Security-sensitive environments
- Academic computing clusters
- Environments without systemd or FUSE support

Apptainer was specifically designed for these use cases and provides:
- **Rootless container execution** (no sudo needed)
- **No Docker daemon required** - pulls directly from registries
- **Works without systemd/FUSE** - direct exec mode
- Better security isolation for multi-tenant systems
- Native HPC integration
- Full compatibility with Docker images

## Technical Implementation Notes

### Exec Mode vs Instance Mode
Initially implemented using Apptainer instance mode (`apptainer instance start`), but discovered this requires systemd and/or FUSE which may not be available in all environments. Switched to direct execution mode (`apptainer exec`) which:
- Works in any environment
- Simpler lifecycle management
- Direct process control (can kill directly)
- No dependency on systemd/FUSE

### Authentication Flow
ApptainerWorkspace discovers SESSION_API_KEY from environment and passes it to RemoteWorkspace, which now properly includes it in the HTTP client's default headers. This ensures all API requests (including conversation creation) are properly authenticated.

## Demo Log

See [`apptainer_workspace_demo.log`](./apptainer_workspace_demo.log) for the complete end-to-end test output showing:
- ✅ Apptainer 1.3.5 detected
- ✅ Image pull without Docker (using cached SIF)
- ✅ Container startup with uvicorn server
- ✅ Health endpoint responding
- ✅ Bash command execution succeeding
- ✅ Conversation creation with authentication
- ✅ Agent initialization and tool setup

## Checklist

- [x] Implementation follows the same pattern as `DockerWorkspace`
- [x] Full RemoteWorkspace API compatibility
- [x] Comprehensive documentation and README
- [x] Usage example included and tested
- [x] Test suite added and passing
- [x] Pre-commit hooks passing
- [x] Type hints and Pydantic validation
- [x] Proper error handling and logging
- [x] Demo log included showing full functionality
- [x] **Tested end-to-end with actual Apptainer 1.3.5** ✨
- [x] **Verified Docker-free operation** ✨
- [x] **Verified systemd/FUSE-free operation** ✨
- [x] **Verified authentication working** ✨

## Next Steps

After merging, users can:
1. Install Apptainer on their systems
2. Use `ApptainerWorkspace` as a drop-in replacement for `DockerWorkspace`
3. Run the example to see it in action
4. Deploy in HPC environments without root access, Docker, systemd, or FUSE
























<!-- AGENT_SERVER_IMAGES_START -->
---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |


**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:1109833-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-1109833-python \
  ghcr.io/openhands/agent-server:1109833-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:1109833-golang-amd64
ghcr.io/openhands/agent-server:1109833-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:1109833-golang-arm64
ghcr.io/openhands/agent-server:1109833-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:1109833-java-amd64
ghcr.io/openhands/agent-server:1109833-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:1109833-java-arm64
ghcr.io/openhands/agent-server:1109833-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:1109833-python-amd64
ghcr.io/openhands/agent-server:1109833-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-amd64
ghcr.io/openhands/agent-server:1109833-python-arm64
ghcr.io/openhands/agent-server:1109833-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-arm64
ghcr.io/openhands/agent-server:1109833-golang
ghcr.io/openhands/agent-server:1109833-java
ghcr.io/openhands/agent-server:1109833-python
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `1109833-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `1109833-python-amd64`) are also available if needed
<!-- AGENT_SERVER_IMAGES_END -->

neubig commented:
## ℹ️ Note on `check-examples` CI Failure

The `check-examples` test is currently failing because it checks that all example files are documented in the [OpenHands/docs](https://github.com/OpenHands/docs) repository. This PR adds a new example file:

- `examples/02_remote_agent_server/05_convo_with_apptainer_sandboxed_server.py`

This example file demonstrates the new `ApptainerWorkspace` functionality and needs to be documented in the docs repository. Once documentation is added there, this check will pass.

### Action Items
- [ ] Add documentation for the Apptainer workspace example in the docs repository
- [ ] Document the new `ApptainerWorkspace` class and its usage

All other CI checks are passing ✅

neubig commented:
@adityasoni9998 you were interested in this, could you try installing apptainer and running the example and see if it works properly? I think it should.

adityasoni9998 commented:
I have been able to successfully run the example `examples/02_remote_agent_server/05_convo_with_apptainer_sandboxed_server.py` only after you merged main into this branch (probably because the docker image had a latest agent server implementation and the client from the older code in this branch was incompatible with the server). 

A small problem I observe is that when I run this example, the TUI shows that the python interpreter is from my local machine though that should not be the case since apptainer has its own python env I guess?

adityasoni9998 commented:
There is another super minor issue that may not be a bug but looks different than how docker runtimes used to behave. It seems that the owner of the files inside the remote apptainer workspace is copied from the username of the local machine I guess? For example, when I run ls -la inside the workspace using `workspace.execute_command`, it shows my Babel username `adityabs` as the file owner. This may possibly not be ideal for reproducibility of results especially when we train smaller models I guess?

<img width="1739" height="723" alt="Screenshot from 2025-12-19 16-31-36" src="https://github.com/user-attachments/assets/1aaf6bd9-2185-4792-9352-e85e818e6920" />

neubig commented:
@openhands install apptainer locally and confirm+reproduce both of the issues pointed out by adityasoni9998. Write the logs demonstrating this issue to disk and tell me the location. Once you have done this, implement a fix.

xingyaoww commented on openhands-workspace/openhands/workspace/apptainer/workspace.py:
```diff
@@ -0,0 +1,403 @@
+        if check_port_available(port):
+            return port
+    return -1
+
```
I believe we already have util function like this in this repo, please re-use those instead of repeatly define it. @OpenHands can you help find the relavant function name and how to import it? Do not edit the code yet

xingyaoww commented on tests/workspace/test_apptainer_workspace.py:
Looks like this example was not launching anything real on apptainer, rather we are mocking everything? Maybe let's remove it?

neubig commented:
@openhands take a look at xingyaoww's comments and take a first pass at resolving them

neubig commented:
Addressed all review comments from @xingyaoww:

1. **Deleted `apptainer-tests.yml`** - Removed the separate workflow since the mocked tests were not testing real functionality
2. **Renamed example to `07_*`** - Changed from `05_convo_with_apptainer_sandboxed_server.py` to `07_convo_with_apptainer_sandboxed_server.py` since example 06 already exists
3. **Removed unnecessary code** - Deleted the `conversation.state._cached_state = None` line from the example
4. **Reused existing util functions** - Now imports `check_port_available` and `find_available_tcp_port` from `docker/workspace.py` instead of duplicating
5. **Removed mocked test file** - Deleted `tests/workspace/test_apptainer_workspace.py` since it was mocking everything
6. **Added setup-apptainer to `run-examples.yml`** - Added `eWaterCycle/setup-apptainer@v2` step so the apptainer example can run in CI
7. **Updated docs PR** - Updated OpenHands/docs#160 to reference the renamed example file (05 → 07)

xingyaoww commented:
@OpenHands download artifact and help me understand why these example test failed

xingyaoww reviewed (approved):
LGTM since example test is passing (the failed ones are not related to apptainer)

xingyaoww commented on openhands-workspace/openhands/workspace/apptainer/README.md:
```diff
@@ -0,0 +1,173 @@
+    host_port=8010,
+) as workspace:
+    result = workspace.execute_command("python --version")
+    print(result.stdout)
```
Actually i think this is not implemented yet- shall we remove `base_image` from the `ApptainerWorkspace`?
